DESIGN.md

```diff
@@ -0,0 +1,55 @@
+# Design
+
+This document covers the design thoughts behind build images for speeding up Travis CI and a proposal for how to implement things going forward.
+
+## Background
+
+Travis CI currently uses standard Ubuntu containers to build curl. There are lots and lots of entries in the curl build matrix and currently a single job takes a long time to complete.
+
+Part of this is due to downloading and compiling dependencies (e.g. brotli, wolfssl). There's also elapsed time taken in downloading and installing packages for compilation.
+
+Using Docker containers, we can create a build image whereby the dependencies are already downloaded and installed, and compiled where necessary. Theoretically it also means developers can have the same build environment as the CI jobs, which is good for reproducibility.
+
+## Design thoughts
+
+- The build images should be based on Ubuntu Trusty, as Travis CI currently use Ubuntu Trusty as their base image.
+  - An alternative could be Centos - any Linux distribution is likely fine. It might be that we choose Trusty to start with then, then move onto Centos for testing different distributions.
+  - Another alternative is Xenial as many of the builds explicitly request the Xenial distribution.
+- Build images will be strictly versioned using semver for reproducibility. The "latest" tag will not be used.
```

Comment:
we have debated 'latest' tag at my company - at best it is something to be considered for production/consumption ... here is a useful link calling it out as an antipattern https://vsupalov.com/docker-latest-tag/

Reply:
Given this is all basically for the CI of curl, I don't believe that we should support :latest for these images. There's no upside. 95% of the time people won't even be using these images (though they can if they want!)